sdk/core/Azure.Core.TestFramework/src/UseSyncMethodsInterceptor.cs

```diff
@@ -186,31 +193,42 @@ private static MethodInfo GetMethod(IInvocation invocation, string nonAsyncMetho
-                    // GetAsync<Model>(Wrapper<Model>) we want that to match
-                    // with Get<T>(Wrapper<T>)
-                    .Where(m =>
-                        m.GetParameters().Select(p => GenericDef(p.ParameterType))
-                        .SequenceEqual(types.Select(GenericDef)))
-                    // Check if they have the same number of type arguments
-                    // (all of our cases today either specialize on 0 or 1 type
-                    // argument for the static or dynamic user schema approach)
-                    .Where(m =>
+            static bool AssignableFromGenericOrEqual(Type t1, Type t2) => t1.IsGenericParameter ? t1.GetGenericParameterConstraints().Any(c => c.IsAssignableFrom(t2)) : t1 == t2;// t.IsAssignableFrom(t2);
+            MethodInfo GetMethodSlow()
+            {
+                var methods = invocation.TargetType
+                // Start with all methods that have the right name
+                .GetMethods(flags).Where(m => m.Name == nonAsyncMethodName);
+                // Check if their type parameters have the same generic
+                // type definitions (i.e., if I invoked
+                // GetAsync<Model>(Wrapper<Model>) we want that to match
+                // with Get<T>(Wrapper<T>)
+                var genericDefs = methods.Where(m =>
+                    m.GetParameters().Select(p => GenericDef(p.ParameterType))
+                    .SequenceEqual(types.Select(GenericDef)));
+
+                // If the previous check has any results, check if they have the same number of type arguments
+                // (all of our cases today either specialize on 0 or 1 type
+                // argument for the static or dynamic user schema approach)
+                // Else, take the method list and filter to one where all args are assignable from their generic constraints
+                // or the types are equal.
+                var withSimilarGenericArguments = genericDefs.Any() ?
+                    genericDefs.Where(m =>
                         m.GetGenericArguments().Length ==
-                        invocation.Method.GetGenericArguments().Length)
-                    // Hopefully we're down to 1.  If you arrive here in the
-                    // future because SingleOrDefault threw, we need to make
-                    // the comparison logic more specific.  If you arrive here
-                    // because we're returning null, then we need to search
-                    // a little more broadly.  Either way, congratulations on
-                    // blazing new API patterns and taking us boldly into the
-                    // future.
-                    .SingleOrDefault();
+                        invocation.Method.GetGenericArguments().Length) :
+                    methods.Where(m =>
+                        m.GetParameters().All(p => AssignableFromGenericOrEqual(p.ParameterType, types[p.Position])));
```

Comment:
In my scenario, the result of the `GenericDef` checks resulted in an empty set, rather than being ambiguous. This `AssignableFromGenericOrEqual` check is one way I found to disambiguate in this scenario, but feedback is welcome on if there is a better way.